### PR TITLE
Initial implementation for Governance Workflows

### PR DESCRIPTION
**Pull request by IceS2**
# Initial Governance Workflows Implementation

The initial implementation focuses on the foundation to build customizable workflows:
- It defines the JSONSchema for Workflow Definition, Instance and Instance State
- It defines the JSONSchema for the Workflow Elements (Nodes, Edges and Triggers)
- Creates the needed Migrations for using Flowable and for te Workflow Entities.
- Creates the needed Endpoints
- Migrates and updates the Glossary Approval Workflow
- Creates the first iteration of the Table Certification Workflow

### Workflow Definition
Defines a Workflow. It is composed of Nodes, Edges and ONE Trigger

### Workflow Instance (Time Series)
It contains all the Workflow Runs (Instances)

### Workflow Instance State (Time Series)
It contains all the Workflow States for each instance.

---

### Edges
Connects two different Nodes. It could contain a condition that branches the flow.

### Nodes

- **Start Event**
Start of the Workflow

- **End Event**
End of the Workflow

- **User Approval Task**
Node that creates a User Task for Approval.
It will be iterated upon once we refactor tasks and have Task Templates.

- **Check Entity Attributes**
Node that allows the user to automatically check for different conditions on the Entity Attributes.

- **Set Entity Certification**
Sets an Entity Certification attribute.

- **Set Glossary Term Status**
Sets a Glossary Term Status.

### Triggers
Configuration needed to actually start the Workflow.

- **Event Based Entity Trigger**
Listens to `Create` and/or `Update` events for a given Entity Type. 

- **Periodic Batch Entity Trygger**
Runs on a given schedule for a given Entity Type.
---

#
### Type of change:

- [x] New feature

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/developers/contribute) document.
- [ ] My PR title is `Fixes <issue-number>: <short explanation>`
- [ ] I have commented on my code, particularly in hard-to-understand areas. 
- [x] For JSON Schema changes: I updated the migration scripts or explained why it is not needed.

### CODE REVIEW

File: bootstrap/sql/migrations/native/1.6.0/mysql/schemaChanges.sql
Comment (TeddyCr) [inline bootstrap/sql/migrations/native/1.6.0/mysql/schemaChanges.sql:1]: Just curious what are the ACT anf FLW tables?

File: openmetadata-service/src/main/java/org/openmetadata/service/governance/workflows/MainWorkflowTerminationListener.java
Comment (TeddyCr) [inline openmetadata-service/src/main/java/org/openmetadata/service/governance/workflows/MainWorkflowTerminationListener.java:1]: What is the difference between Finished and Termination Listener?

Comment (IceS2) [reply openmetadata-service/src/main/java/org/openmetadata/service/governance/workflows/MainWorkflowTerminationListener.java:1]: Termination occurs when the Workflow is forcefully terminated.

File: openmetadata-service/src/main/java/org/openmetadata/service/governance/workflows/elements/Edge.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    SequenceFlow edge = new SequenceFlow(edgeDefinition.getFrom(), edgeDefinition.getTo());
+    if (!CommonUtil.nullOrEmpty(edgeDefinition.getCondition())) {
+      edge.setConditionExpression(getFlowableCondition(edgeDefinition.getCondition()));
+    }
+    this.edge = edge;
+  }
+
+  private String getFlowableCondition(boolean condition) {
+    if (condition) {
+      return "${result}";
```
Comment (TeddyCr) [inline openmetadata-service/src/main/java/org/openmetadata/service/governance/workflows/elements/Edge.java:21]: These would be the JsonLogic expressions?

File: openmetadata-service/src/main/java/org/openmetadata/service/governance/workflows/elements/Edge.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    } else {
+      return "${!result}";
+    }
```
Comment (TeddyCr) [inline openmetadata-service/src/main/java/org/openmetadata/service/governance/workflows/elements/Edge.java:24]: Why do we return `${result}` or `${!result}`; how is the condition used?

File: openmetadata-service/src/main/java/org/openmetadata/service/governance/workflows/elements/nodes/automatedTask/SetEntityCertificationTask.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    attachWorkflowInstanceStageListeners(subProcess);
+
+    this.subProcess = subProcess;
+  }
+
+  private ServiceTask getSetEntityCertificationServiceTask(
+      String subProcessId, String certification) {
+    FieldExtension certificationExpr =
+        new FieldExtensionBuilder()
+            .fieldName("certificationExpr")
```
Comment (TeddyCr) [inline openmetadata-service/src/main/java/org/openmetadata/service/governance/workflows/elements/nodes/automatedTask/SetEntityCertificationTask.java:57]: nit: should we make `certificationExpr` a constant?

File: openmetadata-service/src/main/java/org/openmetadata/service/governance/workflows/elements/nodes/automatedTask/impl/SetEntityCertificationImpl.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    System.out.printf("Set Entity Certification to '%s'%n", certification);
+    //        String originalJson = JsonUtils.pojoToJson(entity);
+    //
+    //        entity.setCertification(certification);
+    //        String updatedJson = JsonUtils.pojoToJson(entity);
+    //
+    //        JsonPatch patch = JsonUtils.getJsonPatch(originalJson, updatedJson);
+    //
+    //        EntityRepository<?> entityRepository = Entity.getEntityRepository(entityType);
+    //        entityRepository.patch(null, entity.getId(), user, patch);
```
Comment (TeddyCr) [inline openmetadata-service/src/main/java/org/openmetadata/service/governance/workflows/elements/nodes/automatedTask/impl/SetEntityCertificationImpl.java:47]: I think we can delete this.

File: openmetadata-service/src/main/java/org/openmetadata/service/governance/workflows/elements/nodes/automatedTask/SetEntityCertificationTask.java (hunk 3 above)
Comment (IceS2) [reply openmetadata-service/src/main/java/org/openmetadata/service/governance/workflows/elements/nodes/automatedTask/SetEntityCertificationTask.java:57]: Hmm... Those strings are only used at that specific place if I'm not mistaken, I've turned a few of the commonly used strings into constant but I think I didn't do that with the Expr ones.

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/CollectionDAO.java [hunk 5]
```diff
@@ -5516,4 +5526,42 @@ default String getNameHashColumn() {
...
+      return "workflow_instance_state_time_series";
+    }
+
+    @SqlQuery(
+        value =
+            "SELECT json FROM workflow_instance_state_time_series "
+                + "WHERE entityFQNHash = :entityFQNHash ORDER BY timestamp DESC")
+    List<String> listWorkflowInstanceStatesForWorkflowInstanceId(
+        @Bind("entityFQNHash") String entityFQNHash);
+  }
```
Comment (TeddyCr) [inline openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/CollectionDAO.java:5566]: Do you think we can use `listWithOffset` from `EntityTimeSeriesDAO` passing the entittyFQNHash as the searchFilter? This way we handle pagination as well.

Comment (IceS2) [reply openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/CollectionDAO.java:5566]: I'll take a look!

File: openmetadata-service/src/main/java/org/openmetadata/service/search/elasticsearch/ElasticSearchClient.java [hunk 6]
```diff
@@ -691,6 +691,72 @@ public SearchResultListMapper listWithOffset(
     }
   }
 
+  @Override
+  public SearchResultListMapper listWithDeepPagination(
```
Comment (TeddyCr) [inline openmetadata-service/src/main/java/org/openmetadata/service/search/elasticsearch/ElasticSearchClient.java:695]: Do you know if we can use `listWithOffset` from the searchRepository?

**Comment by TeddyCr** [comment]
@IceS2 very interesting PR 😊 Are we able to add tests for the workflow buiders pieces?

File: openmetadata-service/src/main/java/org/openmetadata/service/search/elasticsearch/ElasticSearchClient.java (hunk 6 above)
Comment (IceS2) [reply openmetadata-service/src/main/java/org/openmetadata/service/search/elasticsearch/ElasticSearchClient.java:695]: The issue with listWithOffset is that it limits to 10.000 results:
> By default, you cannot use from and size to page through more than 10,000 hits. This limit is a safeguard set by the [index.max_result_window](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules.html#index-max-result-window) index setting. If you need to page through more than 10,000 hits, use the [search_after](https://www.elastic.co/guide/en/elasticsearch/reference/current/paginate-search-results.html#search-after) parameter instead.

**Comment by IceS2** [comment]
@TeddyCr 
We are going to add the tests we are able to once it's in Sandbox Beta due to actually my lack of experience both with Java and with Flowable to actually write them on time 😓

File: openmetadata-service/src/main/java/org/openmetadata/service/governance/workflows/elements/Edge.java (hunk 2 above)
Comment (IceS2) [reply openmetadata-service/src/main/java/org/openmetadata/service/governance/workflows/elements/Edge.java:24]: It is due to how we define Flowable inner logic. For now it is fairly simple and we're just using Booleans. Eventually we need to sort out how to deal with Inputs/Outpus better.

In a nutshell, `result` is the variable we are using. This should be done through constants for sure!

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/CollectionDAO.java (hunk 5 above)
Comment (IceS2) [reply openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/CollectionDAO.java]: Done

File: openmetadata-service/src/main/java/org/openmetadata/service/search/elasticsearch/ElasticSearchClient.java (hunk 6 above)
Comment (TeddyCr) [reply openmetadata-service/src/main/java/org/openmetadata/service/search/elasticsearch/ElasticSearchClient.java:695]: Make sense. Should we change it in `listWithOffset` to use search_after, I think other calls using the same endpoint may end up facing the same issue with the 10K limit --

File: bootstrap/sql/migrations/native/1.6.0/mysql/schemaChanges.sql [hunk 7]
```diff
@@ -1,15 +1,1675 @@
...
+    primary key (ID_)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;
+
+create table ACT_ID_PRIV (
+    ID_ varchar(64) not null,
+    NAME_ varchar(255) not null,
+    primary key (ID_)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;
+
+create table ACT_ID_PRIV_MAPPING (
```
Comment (pmbrull) [inline bootstrap/sql/migrations/native/1.6.0/mysql/schemaChanges.sql:1163]: can't we let flowable run these in a java-based migration? There's a ton of SQL here

File: openmetadata-service/src/main/java/org/openmetadata/service/governance/workflows/WorkflowHandler.java [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+  private final RepositoryService repositoryService;
+  private final RuntimeService runtimeService;
+  private final TaskService taskService;
+  private final HistoryService historyService;
+  private static WorkflowHandler instance;
+  private static volatile boolean initialized = false;
+
+  private WorkflowHandler(OpenMetadataApplicationConfig config) {
+    ProcessEngineConfiguration processEngineConfiguration =
+        new StandaloneProcessEngineConfiguration()
```
Comment (pmbrull) [inline openmetadata-service/src/main/java/org/openmetadata/service/governance/workflows/WorkflowHandler.java:41]: related to the above, maybe this object has some migration utilities to leverage

File: openmetadata-service/src/main/resources/json/data/eventsubscription/WorkflowEvents.json [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+  "fullyQualifiedName": "WorkflowEventConsumer",
+  "displayName": "Workflow Event Consumer",
+  "description": "Consumers EntityChange Events in order to trigger Workflows, if they exist.",
+  "alertType": "GovernanceWorkflowChangeEvent",
+  "filteringRules":{
+    "resources": ["all"]
+  },
+  "destinations": [
+    {
+      "id": "fc9e7a84-5dbd-4e63-8b78-6c3a7bf04a60",
```
Comment (pmbrull) [inline openmetadata-service/src/main/resources/json/data/eventsubscription/WorkflowEvents.json:12]: what's this ID?

File: openmetadata-spec/src/main/java/org/openmetadata/schema/governance/workflows/elements/WorkflowNodeDefinitionInterface.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+  default Object getConfig() {
+    return null;
+  }
+  ;
+
+  default List<String> getInputs() {
+    return null;
+  }
+  ;
```
Comment (pmbrull) [inline openmetadata-spec/src/main/java/org/openmetadata/schema/governance/workflows/elements/WorkflowNodeDefinitionInterface.java:27]: nit - do we need the ; here?

File: bootstrap/sql/migrations/native/1.6.0/mysql/schemaChanges.sql (hunk 7 above)
Comment (IceS2) [reply bootstrap/sql/migrations/native/1.6.0/mysql/schemaChanges.sql:1163]: I discusses this with Mohit. Technically Flowable can automatically take care of the migrations, but we decided to do it like we did with quartz, manually.

Curiously enough, I had to change the default charset that they set to `utf8mb4` from `utf8`, so I actually already modified them

File: openmetadata-service/src/main/java/org/openmetadata/service/governance/workflows/WorkflowHandler.java (hunk 8 above)
Comment (IceS2) [reply openmetadata-service/src/main/java/org/openmetadata/service/governance/workflows/WorkflowHandler.java:41]: Not sure I understood the comment for this specific configuration

File: openmetadata-service/src/main/resources/json/data/eventsubscription/WorkflowEvents.json (hunk 9 above)
Comment (IceS2) [reply openmetadata-service/src/main/resources/json/data/eventsubscription/WorkflowEvents.json:12]: Literally just a randomly generated ID that is used for this seed data

File: openmetadata-ui/src/main/resources/ui/src/components/AppRouter/withAdvanceSearch.tsx [hunk 11]
```diff
@@ -10,18 +10,31 @@
...
+  SearchOutputType,
+} from '../Explore/AdvanceSearchProvider/AdvanceSearchProvider.interface';
 
 export const withAdvanceSearch =
   <P extends Record<string, unknown>>(
     Component: FC<P>,
     providerProps?: Omit<AdvanceSearchProviderProps, 'children'>
   ) =>
   (props: P) => {
+    // console.log('providerProps', props);
```
Comment (chirag-madlani) [inline openmetadata-ui/src/main/resources/ui/src/components/AppRouter/withAdvanceSearch.tsx:27]: nit

**Review by pmbrull — APPROVED** [review]

**Review by pmbrull — APPROVED** [review]

**Review by pmbrull — DISMISSED** [review]

**Review by pmbrull — APPROVED** [review]

**Pull request merged**